ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureStorageContainerMetricsCollector.java

```diff
@@ -0,0 +1,112 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A class that aggregates container metrics. We do _NOT_ want to emit per-container metrics because then the number
+ * of metrics increases proportionally overwhelming telemetry. This was attempted and failed.
+ *
+ * A daemon will run at regular intervals emitting aggregate metrics in a controlled and predictable manner.
+ * This is a singleton class to avoid multiple collector threads.
+ */
+public class AzureStorageContainerMetricsCollector {
+  private final AzureMetrics metrics;
+  private final ConcurrentHashMap<Long, AzureStorageContainerMetrics> metricMap;
+  private final ScheduledExecutorService executor;
+  private static AzureStorageContainerMetricsCollector instance;
+  private final VerifiableProperties properties;
+
+  public static final Logger logger = LoggerFactory.getLogger(AzureStorageContainerMetricsCollector.class);
+
+  private AzureStorageContainerMetricsCollector(MetricRegistry metrics, VerifiableProperties properties) {
+    metricMap = new ConcurrentHashMap<>();
+    this.metrics = new AzureMetrics(metrics);
+    this.properties = properties;
+    executor = Utils.newScheduler(1, "azure_storage_container_metrics_collector_", true);
+    executor.scheduleWithFixedDelay(getCollector(), 0, 2, TimeUnit.MINUTES);
+    logger.info("Started AzureStorageContainerMetricsCollector");
+  }
+
+  private Runnable getCollector() {
+    return () -> {
+      Long totalLag = metricMap.values().stream()
+          .map(container -> container.getPartitionLag())
+          .reduce(0L, Long::sum);
+      this.metrics.azureContainerLagBytesCount.inc(totalLag);
+    };
+  }
+
+  /**
+   * Thread-safe singleton initializer
+   * @param metrics
+   * @return collector instance
+   */
+  public static synchronized AzureStorageContainerMetricsCollector getInstance(MetricRegistry metrics,
+      VerifiableProperties properties) {
+    if (instance == null) {
+      instance = new AzureStorageContainerMetricsCollector(metrics, properties);
+    }
+    return instance;
+  }
+
+  public void addPartitionReplicas(List<RemoteReplicaInfo> remoteReplicaInfos) {
+    for (RemoteReplicaInfo rinfo : remoteReplicaInfos) {
+      // Don't store any references to PartitionId or RemoteReplicaInfo.
+      // With improper clean up, these references linger around and cause memory leaks.
+      long pid = rinfo.getReplicaId().getPartitionId().toPathString().hashCode();
+      metricMap.putIfAbsent(pid, new AzureStorageContainerMetrics(pid));
+      metricMap.get(pid).addPartitionReplica(rinfo.getReplicaId().getDataNodeId().getHostname());
+    }
+  }
+
+  public void removePartitionReplicas(List<RemoteReplicaInfo> remoteReplicaInfos) {
+    for (RemoteReplicaInfo rinfo : remoteReplicaInfos) {
+      long pid = rinfo.getReplicaId().getPartitionId().toPathString().hashCode();
```

Comment:
let's not use hashCode here, just call `rinfo.getReplicaId().getPartitionId().getId()`. You are already calling it from VcrReplicaThread.

Reply:
oh good catch, that's a typo !